libioc/Provisioning/puppet.py

```diff
@@ -0,0 +1,233 @@
+        """Return the Puppet Control-Repo URL."""
+        return str(self._url)
+
+    @url.setter
+    def url(self, value: str) -> None:
+        """Set the Puppet Control-Repo URL."""
+        self._url = value
+
+    @property
+    def name(self) -> str:
+        """Return the unique name for this Puppet Control-Repo URL."""
+        return self._name
+
+    @name.setter
+    def name(self, value: str) -> None:
+        """Set a unique name for this Puppet Control-Repo URL."""
+        self._name = value
+
+    @property
+    def pkgs(self) -> typing.List[str]:
+        """Return list of packages required for this Provisioning method."""
+        return self._pkgs
+
+    def generate_postinstall(self) -> str:
+        """Return list of strings representing our postinstall."""
+        postinstall = """#!/bin/sh
+        set -eu
+
+        """
+
+        if self.remote:
+            postinstall += """cat > /usr/local/etc/r10k/r10k.yml <EOF
+            ---
+            :source:
+                puppet:
+                    basedir: /usr/local/etc/puppet/environments
+                    remote: {self.url}
+            >EOF
+
+            r10k deploy environment -p
+
+            """
+
+        postinstall += """
+        puppet apply /usr/local/etc/puppet/environments/manifests/site.pp
+        """
+        return postinstall
+
+
+def provision(
+    self: 'libioc.Provisioning.Prototype',
+    event_scope: typing.Optional['libioc.events.Scope']=None
+) -> typing.Generator['libioc.events.IocEvent', None, None]:
+    r"""
+    Provision the jail with Puppet apply using the supplied control-repo.
+
+    The repo can either be a filesystem path, or a http[s]/git URL.
+    If the repo is a filesystem path, it will be mounted appropriately.
+    If the repo is a URL, it will be setup with `r10k`.
+
+        ioc set \
+            provisioning.method=puppet \
+            provisioning.source=http://example.com/my/puppet-env \
+            provisioning.name=my-puppet-env \
+            myjail
+
+    """
+    events = libioc.events
+    jailProvisioningEvent = events.JailProvisioning(
+        jail=self.jail,
+        scope=event_scope
+    )
+    yield jailProvisioningEvent.begin()
+    _scope = jailProvisioningEvent.scope
+    jailProvisioningAssetDownloadEvent = events.JailProvisioningAssetDownload(
+        jail=self.jail,
+        scope=_scope
+    )
+
+    # download / mount provisioning assets
+    try:
+        yield jailProvisioningAssetDownloadEvent.begin()
+        pluginDefinition = ControlRepoDefinition(
+            name=self.name,
+            url=self.source,
+            logger=self.jail.logger
+        )
+        yield jailProvisioningAssetDownloadEvent.end()
+    except Exception as e:
+        yield jailProvisioningAssetDownloadEvent.fail(e)
+        raise e
+
+    if pluginDefinition.remote:
+        mode = 'rw'  # we'll need to run r10k here..
+        plugin_dataset_name = f"{self.jail.dataset.name}/puppet"
```

Comment:
given the effort we've put into allowing `provisioning.name`, it would be good to download plugins under their own dataset